File: java/src/main/java/br/com/vps/consulting/b2b/management/partner/infrastructure/persistence/DefaultPartnerRepository.java
package br.com.vps.consulting.b2b.management.partner.infrastructure.persistence;

import br.com.vps.consulting.b2b.management.partner.domain.Partner;
import br.com.vps.consulting.b2b.management.partner.domain.PartnerCredit;
import br.com.vps.consulting.b2b.management.partner.domain.PartnerId;
import br.com.vps.consulting.b2b.management.partner.domain.PartnerRepository;
import br.com.vps.consulting.b2b.management.partner.infrastructure.mapper.PartnerCreditMapper;
import br.com.vps.consulting.b2b.management.partner.infrastructure.mapper.PartnerMapper;
import br.com.vps.consulting.b2b.management.partner.infrastructure.persistence.jpa.PartnerCreditJpaRepository;
import br.com.vps.consulting.b2b.management.partner.infrastructure.persistence.jpa.PartnerJpaRepository;
import br.com.vps.consulting.b2b.management.shared.core.page.PageCustom;
import jakarta.inject.Named;
import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;
import org.springframework.data.domain.Sort;
import org.springframework.stereotype.Component;

import java.math.BigDecimal;
import java.util.Optional;

@Component
@RequiredArgsConstructor
public class DefaultPartnerRepository implements PartnerRepository {

    private final PartnerJpaRepository partnerJpaRepository;
    private final PartnerCreditJpaRepository partnerCreditJpaRepository;

    @Override
    public Partner save(final Partner partner) {
        final var partnerEntity = PartnerMapper.toEntity(partner);
        partnerJpaRepository.save(partnerEntity);
        partnerCreditJpaRepository.save(PartnerCreditMapper.toEntity(partner));
        return partner;
    }

    @Override
    public Optional<Partner> findById(final PartnerId id) {
        return partnerJpaRepository.findById(id.value())
                .map(partnerEntity -> {
                    final var credit = partnerCreditJpaRepository.findById(id.value()).orElse(null);
                    return PartnerMapper.toDomain(partnerEntity, credit);
                });
    }

    @Override
    public Optional<PartnerCredit> findCreditById(final PartnerId id) {
        return partnerCreditJpaRepository.findById(id.value())
                .map(e -> new PartnerCredit(e.getCreditLimit(), e.getAvailableBalance(), e.getReservedBalance(), e.getUpdatedAt()));
    }

    @Override
    public boolean reserveCredit(final PartnerId partnerId, final BigDecimal amount) {
        return partnerCreditJpaRepository.reserveCredit(partnerId.value(), amount) > 0;
    }

    @Override
    public void debitReservation(final PartnerId partnerId, final BigDecimal amount) {
        partnerCreditJpaRepository.debitReservation(partnerId.value(), amount);
    }

    @Override
    public void releaseReservation(final PartnerId partnerId, final BigDecimal amount) {
        partnerCreditJpaRepository.releaseReservation(partnerId.value(), amount);
    }

    @Override
    public void refundCredit(final PartnerId partnerId, final BigDecimal amount) {
        partnerCreditJpaRepository.refundCredit(partnerId.value(), amount);
    }

    @Override
    public void adjustCreditLimit(final PartnerId partnerId, final BigDecimal newLimit) {
        partnerCreditJpaRepository.adjustCreditLimit(partnerId.value(), newLimit);
    }

    @Override
    public PageCustom<Partner> findAll(final long pageSize, final long pageNumber) {
        final Pageable pageable = PageRequest.of((int) pageNumber, (int) pageSize,
                Sort.by(Sort.Direction.DESC, "createdAt"));
        final var result = partnerJpaRepository.findAll(pageable);
        return PageCustom.<Partner>builder()
                .pageNumber(result.getNumber())
                .pageSize(result.getSize())
                .numberOfElements(result.getNumberOfElements())
                .totalPages(result.getTotalPages())
                .totalElements(result.getTotalElements())
                .items(result.getContent().stream().map(e -> PartnerMapper.toDomain(e, null)).toList())
                .build();
    }

}

File: kotlin/src/main/kotlin/br/com/vps/consulting/b2b/management/partner/infrastructure/persistence/DefaultPartnerRepository.kt
package br.com.vps.consulting.b2b.management.partner.infrastructure.persistence

import br.com.vps.consulting.b2b.management.partner.domain.Partner
import br.com.vps.consulting.b2b.management.partner.domain.PartnerBalance
import br.com.vps.consulting.b2b.management.partner.domain.PartnerId
import br.com.vps.consulting.b2b.management.partner.domain.PartnerRepository
import br.com.vps.consulting.b2b.management.partner.infrastructure.mapper.toDomain
import br.com.vps.consulting.b2b.management.partner.infrastructure.mapper.toEntity
import br.com.vps.consulting.b2b.management.partner.infrastructure.mapper.zeroBalanceEntityFor
import br.com.vps.consulting.b2b.management.partner.infrastructure.persistence.jpa.PartnerBalanceJpaRepository
import br.com.vps.consulting.b2b.management.partner.infrastructure.persistence.jpa.PartnerJpaRepository
import br.com.vps.consulting.b2b.management.partner.infrastructure.persistence.jpa.PartnerSpecification
import br.com.vps.consulting.b2b.management.shared.core.page.PageCustom
import br.com.vps.consulting.b2b.management.shared.core.vo.Money
import org.springframework.data.domain.PageRequest
import org.springframework.data.domain.Sort
import org.springframework.data.jpa.domain.Specification
import org.springframework.stereotype.Component

@Component
class DefaultPartnerRepository(
    private val partnerJpaRepository: PartnerJpaRepository,
    private val partnerBalanceJpaRepository: PartnerBalanceJpaRepository,
) : PartnerRepository {

    override fun save(partner: Partner): Partner {
        partnerJpaRepository.save(partner.toEntity())
        partnerBalanceJpaRepository.save(zeroBalanceEntityFor(partner.id.value))
        return partner
    }

    override fun findById(id: PartnerId): Partner? =
        partnerJpaRepository.findById(id.value).map { it.toDomain() }.orElse(null)

    override fun findBalanceById(id: PartnerId): PartnerBalance? =
        partnerBalanceJpaRepository.findById(id.value).map { it.toDomain() }.orElse(null)

    override fun findAll(document: String?, pageSize: Int, pageNumber: Int): PageCustom<Partner> {
        val spec = Specification.where(PartnerSpecification.hasDocument(document))
        val pageable = PageRequest.of(pageNumber, pageSize, Sort.by(Sort.Direction.DESC, "createdAt"))
        val result = partnerJpaRepository.findAll(spec, pageable)
        return PageCustom(
            pageNumber = result.number,
            pageSize = result.size,
            totalPages = result.totalPages,
            totalElements = result.totalElements,
            items = result.content.map { it.toDomain() },
        )
    }

    override fun creditBalance(partnerId: PartnerId, amount: Money): Boolean =
        partnerBalanceJpaRepository.creditBalance(partnerId.value, amount.value) > 0

    override fun debitBalance(partnerId: PartnerId, amount: Money): Boolean =
        partnerBalanceJpaRepository.debitBalance(partnerId.value, amount.value) > 0
}
